### PR TITLE
QPACK security considerations

### LINKED ISSUE #1737: QPACK has no Security Considerations
The full text of that section is "TBD."

### PR DESCRIPTION
**Pull request by afrind**
Behold

Fixes #1737

### CODE REVIEW

File: draft-ietf-quic-qpack.md [hunk 1]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+attacker can observe the encoded length and infer that the guess was correct.
+
+This is possible even over the Transport Layer Security Protocol (TLS, see
+{{?RFC5246}}), because while TLS provides confidentiality protection for
+content, it only provides a limited amount of protection for the length of that
+content.
+
+Note:
+
+Padding schemes only provide limited protection against an attacker with these
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1210]:
```suggestion
: Padding schemes only provide limited protection against an attacker with these
```

File: draft-ietf-quic-qpack.md [hunk 2]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+Rather than impose constraints on users of HTTP, an implementation of QPACK can
+instead constrain how compression is applied in order to limit the potential for
+dynamic table probing.
+
+An ideal solution segregates access to the dynamic table based on the entity
+that is constructing header fields. Header field values that are added to the
+table are attributed to an entity, and only the entity that created a particular
+value can extract that value.
+
+To improve compression performance of tqhis option, certain entries might be tagged as being public. For example, a web browser might make the values of the Accept-Encoding header field available in all requests.
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1266]:
```suggestion
To improve compression performance of this option, certain entries might be
tagged as being public. For example, a web browser might make the values of the
Accept-Encoding header field available in all requests.
```

File: draft-ietf-quic-qpack.md [hunk 3]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+
+An encoder without good knowledge of the provenance of header fields might
+instead introduce a penalty for a header field with many different values, such
+that a large number of attempts to guess a header field value results in the
+header field no more being compared to the dynamic table entries in future
+messages, effectively preventing further guesses.
+
+Note:
+
+Simply removing entries corresponding to the header field from the dynamic table
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1276]:
```suggestion
: Simply removing entries corresponding to the header field from the dynamic table
```

File: draft-ietf-quic-qpack.md [hunk 4]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+table are attributed to an entity, and only the entity that created a particular
+value can extract that value.
+
+To improve compression performance of tqhis option, certain entries might be tagged as being public. For example, a web browser might make the values of the Accept-Encoding header field available in all requests.
+
+An encoder without good knowledge of the provenance of header fields might
+instead introduce a penalty for a header field with many different values, such
+that a large number of attempts to guess a header field value results in the
+header field no more being compared to the dynamic table entries in future
+messages, effectively preventing further guesses.
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1272]:
```suggestion
that a large number of attempts to guess a header field value results in the
header field not being compared to the dynamic table entries in future
messages, effectively preventing further guesses.
```

I had a little trouble with "no more".

File: draft-ietf-quic-qpack.md [hunk 5]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+can be ineffectual if the attacker has a reliable way of causing values to be
+reinstalled. For example, a request to load an image in a web browser typically
+includes the Cookie header field (a potentially highly valued target for this
+sort of attack), and web sites can easily force an image to be loaded, thereby
+refreshing the entry in the dynamic table.
+
+This response might be made inversely proportional to the length of the header
+field value. Marking a header field as not using the dynamic table any more
+might occur for shorter values more quickly or with higher probability than for
+longer values.
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1286]:
```suggestion
field value. Disabling access to the dynamic table for a header field might
occur for shorter values more quickly or with higher probability than for longer
values.
```

File: draft-ietf-quic-qpack.md [hunk 6]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+On the contrary, an encoder might prefer indexing values for header fields that
+have little or no value if they were exposed. For instance, a User-Agent header
+field does not commonly vary between requests and is sent to any server. In that
+case, confirmation that a particular User-Agent value has been used provides
+little value.
+
+Note that these criteria for deciding to use a never indexed literal
+representation will evolve over time as new attacks are discovered.
+
+##. Static Huffman Encoding
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1323]:
```suggestion
## Static Huffman Encoding
```

File: draft-ietf-quic-qpack.md [hunk 7]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+
+There is no currently known attack against a static Huffman encoding. A study
+has shown that using a static Huffman encoding table created an information
+leakage, however this same study concluded that an attacker could not take
+advantage of this information leakage to recover any meaningful amount of
+information (see [PETAL]).
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1329]: Want to point out that QPACK uses the same Huffman table as HPACK?

File: draft-ietf-quic-qpack.md [hunk 8]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+
+## Memory Consumption
+
+An attacker can try to cause an endpoint to exhaust its memory. QPACK is
+designed to limit both the peak and state amounts of memory allocated by an
+endpoint.
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1335]:
```suggestion
designed to limit both the peak and stable amounts of memory allocated by an
endpoint.
```

File: draft-ietf-quic-qpack.md [hunk 9]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+
+The amount of memory used by the compressor is limited by the protocol using
+QPACK through the definition of the maximum size of the dynamic table, and the
+maximum number of blocking streams. In HTTP/3, these values are controlled by
+the decoder through the setting parameter QPACK_MAX_TABLE_CAPACITY and
+QPACK_BLOCKED_STREAMS, respectively (see Section
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1341]:
```suggestion
QPACK_BLOCKED_STREAMS, respectively (see
```

File: draft-ietf-quic-qpack.md [hunk 10]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+of blocked streams is only a proxy for the maximum amount of memory required by
+the decoder.  The actual maximum amount of memory will depend on how much memory
+the decoder uses to track each blocked stream.
+
+A decoder can limit the amount of state memory used for the dynamic table by
+setting an appropriate value for the maximum size of the dynamic table. In
+HTTP/3, this is realized by setting an appropriate value for the
+QPACK_MAX_TABLE_CAPACITY parameter. An encoder can limit the amount of state
+memory it uses by signaling lower dynamic table size than the decoder allows
+(see {{eviction}}).  A decoder can limit the amount of state memory used for
```
Comment (martinthomson) [inline draft-ietf-quic-qpack.md:1354]:
```suggestion
(see {{eviction}}).

A decoder can limit the amount of state memory used for
```

File: draft-ietf-quic-qpack.md [hunk 11]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+by forcing a guess to match an entire header field value, rather than individual
+characters. An attacker can only learn whether a guess is correct or not, so is
+reduced to a brute force guess for the header field values.
+
+The viability of recovering specific header field values therefore depends on
+the entropy of values. As a result, values with high entropy are unlikely to be
+recovered successfully. However, values with low entropy remain vulnerable.
+
+Attacks of this nature are possible any time that two mutually distrustful
+entities control requests or responses that are placed onto a single HTTP/2
```
Comment (LPardue) [inline draft-ietf-quic-qpack.md:1233]:
```suggestion
entities control requests or responses that are placed onto a single HTTP/3
```

File: draft-ietf-quic-qpack.md [hunk 12]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+size of the dynamic table takes into account both the size of the data stored in
+the dynamic table, plus a small allowance for overhead.  The limit on the number
+of blocked streams is only a proxy for the maximum amount of memory required by
+the decoder.  The actual maximum amount of memory will depend on how much memory
+the decoder uses to track each blocked stream.
+
+A decoder can limit the amount of state memory used for the dynamic table by
+setting an appropriate value for the maximum size of the dynamic table. In
+HTTP/3, this is realized by setting an appropriate value for the
+QPACK_MAX_TABLE_CAPACITY parameter. An encoder can limit the amount of state
```
Comment (LPardue) [inline draft-ietf-quic-qpack.md:1352]: nit: the doc is inconsistent about QPACK_MAX_TABLE_CAPACITY  vs SETTINGS_QPACK_MAX_TABLE_CAPACITY. I don't know what the answer is

File: draft-ietf-quic-qpack.md [hunk 13]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+memory it uses by signaling lower dynamic table size than the decoder allows
```
Comment (LPardue) [inline draft-ietf-quic-qpack.md:1353]:
```suggestion
memory it uses by signaling a lower dynamic table size than the decoder allows
```

File: draft-ietf-quic-qpack.md [hunk 14]
```diff
@@ -1144,7 +1179,193 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+(see {{eviction}}).  A decoder can limit the amount of state memory used for
+blocked streams by setting an appropriate value for the maximum number of
+blocked streams.  In HTTP/3, this is realized by setting an appropriate value
+for the QPACK_BLOCKED_STREAMS parameter.  An encoder can limit the amount of
```
Comment (LPardue) [inline draft-ietf-quic-qpack.md:1357]: nit: QPACK_BLOCKED_STREAMS  vs. SETTINGS_QPACK_BLOCKED_STREAMS

File: draft-ietf-quic-qpack.md (hunk 7 above)
Comment (afrind) [reply draft-ietf-quic-qpack.md:1329]: It's mentioned in the string literal section: 

> When Huffman encoding is enabled, the Huffman
table from Appendix B of [RFC7541] is used without modification.

File: draft-ietf-quic-qpack.md (hunk 8 above)
Comment (afrind) [reply draft-ietf-quic-qpack.md:1335]: This error is in the HPACK RFC.  Is it worth an errata?

File: draft-ietf-quic-qpack.md (hunk 12 above)
Comment (afrind) [reply draft-ietf-quic-qpack.md:1352]: The text seems to consistently use SETTINGS_ variants, but the table omits it.  I'll fix these to match the text, but SETTINGS_ seems redundant in the table.  I'll resolve the inconsistency in another PR.

Comment (LPardue) [reply draft-ietf-quic-qpack.md:1352]: wfm

File: draft-ietf-quic-qpack.md (hunk 8 above)
Comment (martinthomson) [reply draft-ietf-quic-qpack.md]: I'll let you manage that :)  You can be the author of an RFC erratum.  The glory!

File: draft-ietf-quic-qpack.md (hunk 12 above)
Comment (MikeBishop) [reply draft-ietf-quic-qpack.md]: It's an (unstated) norm to omit SETTINGS_ in cases where it's already blatantly obvious that we're discussing a setting.  You'll find similar variance in 7540, 7541, and HTTP/3.

File: draft-ietf-quic-qpack.md [hunk 15]
```diff
@@ -1144,7 +1168,197 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+Refusing to insert a header field into the dynamic table is only
+effective if doing so is avoided on all hops. The never indexed literal bit (see
+{{literal-name-reference}}) can be used to signal to intermediaries that a
+particular value was intentionally sent as a literal.
+
+An intermediary MUST NOT re-encode a value that uses a literal representation
+with the 'N' bit set with another representation that would index it. If QPACK
+is used for re-encoding, a literal representation with the 'N' bit set MUST be
+used.  If HPACK (see Section 6.2.3 of [RFC7541]) is used for re-encoding, the
+never indexed literal representation MUST be used.
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md:1293]:
```suggestion
used.  If HPACK is used for re-encoding, the never indexed literal
representation (see Section 6.2.3 of [RFC7541]) MUST be used.
```

File: draft-ietf-quic-qpack.md [hunk 16]
```diff
@@ -1144,7 +1168,197 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+
+The choice to use a never indexed literal representation for a header field
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md:1295]:
```suggestion
The choice to mark that a header field should never be indexed
```

File: draft-ietf-quic-qpack.md [hunk 17]
```diff
@@ -1144,7 +1168,197 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+
+An attacker can try to cause an endpoint to exhaust its memory. QPACK is
+designed to limit both the peak and stable amounts of memory allocated by an
+endpoint.
+
+The amount of memory used by the compressor is limited by the protocol using
+QPACK through the definition of the maximum size of the dynamic table, and the
+maximum number of blocking streams. In HTTP/3, these values are controlled by
+the decoder through the setting parameter SETTINGS_QPACK_MAX_TABLE_CAPACITY and
+SETTINGS_QPACK_BLOCKED_STREAMS, respectively (see
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md:1332]:
```suggestion
the decoder through the settings parameters SETTINGS_QPACK_MAX_TABLE_CAPACITY
and SETTINGS_QPACK_BLOCKED_STREAMS, respectively (see
```

File: draft-ietf-quic-qpack.md [hunk 18]
```diff
@@ -1144,7 +1168,197 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+advantage of this information leakage to recover any meaningful amount of
+information (see [PETAL]).
+
+## Memory Consumption
+
+An attacker can try to cause an endpoint to exhaust its memory. QPACK is
+designed to limit both the peak and stable amounts of memory allocated by an
+endpoint.
+
+The amount of memory used by the compressor is limited by the protocol using
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md:1328]:
```suggestion
The amount of memory used by the encoder is limited by the protocol using
```

File: draft-ietf-quic-qpack.md [hunk 19]
```diff
@@ -1144,7 +1168,197 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+HTTP/3, this is realized by setting an appropriate value for the
+QPACK_BLOCKED_STREAMS parameter.  An encoder can limit the amount of state
+memory by only using as many blocked streams as it wishes to support; no
+signaling to the decoder is requred.
+
+The amount of temporary memory consumed by an encoder or decoder can be limited
+by processing header fields sequentially. A decoder implementation does not need
+to retain a complete list of header fields while decoding a header block. An
+encoder implementation does not need to retain a complete list of header fields
+if while encoding a header block if it is using a single-pass encoder.  Note
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md:1358]:
```suggestion
while encoding a header block if it is using a single-pass algorithm.  Note
```

File: draft-ietf-quic-qpack.md [hunk 20]
```diff
@@ -1155,6 +1369,19 @@ to an excess of unsent data might include limiting the ability of the peer to
...
+## Implementation Limits
+
+An implementation of QPACK needs to ensure that large values for integers, long
+encoding for integers, or long string literals do not create security
+weaknesses.
+
+An implementation has to set a limit for the values it accepts for integers, as
+well as for the encoded length (see {{prefixed-integers}}). In the same way, it
+has to set a limit to the length it accepts for string literals (see
+{{string-literals}}).
```
Comment (MikeBishop) [inline draft-ietf-quic-qpack.md:1382]: I believe we have a minimum size which MUST be supported; should that be mentioned here?

File: draft-ietf-quic-qpack.md [hunk 21]
```diff
@@ -1144,7 +1168,197 @@ QPACK_DECODER_STREAM_ERROR (0x202):
...
+endpoint.
+
+The amount of memory used by the compressor is limited by the protocol using
+QPACK through the definition of the maximum size of the dynamic table, and the
+maximum number of blocking streams. In HTTP/3, these values are controlled by
+the decoder through the setting parameter SETTINGS_QPACK_MAX_TABLE_CAPACITY and
+SETTINGS_QPACK_BLOCKED_STREAMS, respectively (see
+{{maximum-dynamic-table-capacity}} and {{blocked-streams}}). The limit on the
+size of the dynamic table takes into account both the size of the data stored in
+the dynamic table, plus a small allowance for overhead.  The limit on the number
```
Comment (bencebeky) [inline draft-ietf-quic-qpack.md:1335]: "both" does not flow well with "plus".  Either s/", plus"/" and"/, or remove "both".

**Comment by LPardue** [comment]
@afrind I think this is inches of getting landed

File: draft-ietf-quic-qpack.md (hunk 20 above)
Comment (afrind) [reply draft-ietf-quic-qpack.md:1382]: @MikeBishop : I can't quite figure an elegant way to mention the minimum here.  It is already mentioned in the referenced section.

**Pull request merged**